fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/MultidimensionalIndexMaintainer.java

```diff
@@ -0,0 +1,545 @@
+            if (prefixSize > 0) {
+                rtSubspace = indexSubspace.subspace(prefixKey);
+            } else {
+                rtSubspace = indexSubspace;
+            }
+
+            // It is unsafe to have two concurrent updates to the same R-tree, so ensure that at most
+            // one update per prefix key is ongoing at any given time
+            final Function<Void, CompletableFuture<Void>> futureSupplier =
+                    ignored -> {
+                        final RTree.Point point =
+                                validatePoint(new RTree.Point(Tuple.fromList(indexKeyItems.subList(prefixSize, prefixSize + dimensionsSize))));
+
+                        final List<Object> primaryKeyParts = Lists.newArrayList(savedRecord.getPrimaryKey().getItems());
+                        state.index.trimPrimaryKey(primaryKeyParts);
+                        final List<Object> keySuffixParts =
+                                Lists.newArrayList(indexKeyItems.subList(prefixSize + dimensionsSize, indexKeyItems.size()));
+                        keySuffixParts.addAll(primaryKeyParts);
+                        final Tuple keySuffix = Tuple.fromList(keySuffixParts);
+                        final FDBStoreTimer timer = Objects.requireNonNull(getTimer());
+                        final RTree rTree = new RTree(rtSubspace, getExecutor(), config, RTreeHilbertCurveHelpers::hilbertValue,
+                                RTree::newRandomNodeId, new OnWrite(timer), RTree.OnReadListener.NOOP);
+                        if (remove) {
+                            return rTree.delete(state.transaction, point, keySuffix);
+                        } else {
+                            return rTree.insertOrUpdate(state.transaction,
+                                    point,
+                                    keySuffix,
+                                    indexEntry.getValue());
+                        }
+                    };
+            final CompletableFuture<Void> existingFuture = rankFutures.get(rtSubspace);
+            if (existingFuture == null) {
+                rankFutures.put(rtSubspace, futureSupplier.apply(null));
+            } else {
+                rankFutures.put(rtSubspace, existingFuture.thenCompose(futureSupplier));
+            }
+        }
+        return AsyncUtil.whenAll(rankFutures.values());
+    }
+
+    private RTree.Point validatePoint(@Nonnull RTree.Point point) {
+        for (int d = 0; d < point.getNumDimensions(); d ++) {
+            Object coordinate = point.getCoordinate(d);
+            Preconditions.checkArgument(coordinate == null || coordinate instanceof Long,
+                    "dimension coordinates must be of type long");
+        }
+        return point;
+    }
+
+    @Override
+    public CompletableFuture<Void> deleteWhere(Transaction tr, @Nonnull Tuple prefix) {
+        Verify.verify(getDimensionsKeyExpression(state.index.getRootExpression()).getPrefixSize() >= prefix.size());
```

Comment:
I think a precondition error like this is only appropriate if there is a `canDeleteWhere` that checked it (or its equivalent).

Reply:
Done! Also added a test case for `deleteWhere()`